cli: TestZipQueryFallback failed

cli.TestZipQueryFallback [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/21127973?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/21127973?buildTab=artifacts#/) on release-26.1 @ [5122355c4cddec33691f2bedb42feb36b7785070](https://github.com/cockroachdb/cockroach/commits/5122355c4cddec33691f2bedb42feb36b7785070):


```
          [cluster] tenant hot range summary script... writing binary output: debug/hot-ranges-tenant.sh... done
          [cluster] capture debug zip flags... writing binary output: debug/debug_zip_command_flags.txt... done
        output didn't match expected:
          @@ -8,41 +8,8 @@
           [cluster] requesting data for debug/settings... received response... writing JSON output: debug/settings.json... done
           [cluster] requesting data for debug/reports/problemranges... received response... writing JSON output: debug/reports/problemranges.json... done
           <dumping SQL tables>
           [cluster] retrieving SQL data for crdb_internal.transaction_contention_events: last request failed: ERROR: column "fail" does not exist (SQLSTATE 42703)
           <dumping SQL tables>
          -[cluster] requesting nodes... received response... writing JSON output: debug/nodes.json... done
          -[cluster] requesting liveness... received response... writing JSON output: debug/liveness.json... done
          -[cluster] requesting tenant ranges... received response...
          -[cluster] requesting tenant ranges: last request failed: rpc error: ...
          -[cluster] requesting tenant ranges: creating error output: debug/tenant_ranges.err.txt... done
          -[cluster] collecting the inflight traces for jobs... received response... done
          -[cluster] requesting CPU profiles
          -[cluster] profiles generated
          -[cluster] profile for node 1... writing binary output: debug/nodes/1/cpu.pprof... done
          -[node 1] node status... writing JSON output: debug/nodes/1/status.json... done
          -[node 1] using SQL connection URL: postgresql://...
          -<dumping SQL tables>
          -[node 1] requesting data for debug/nodes/1/details... received response... writing JSON output: debug/nodes/1/details.json... done
          -[node 1] requesting data for debug/nodes/1/gossip... received response... writing JSON output: debug/nodes/1/gossip.json... done
          -[node 1] requesting stacks... received response... writing binary output: debug/nodes/1/stacks.txt... done
          -[node 1] requesting stacks with labels... received response... writing binary output: debug/nodes/1/stacks_with_labels.txt... done
          -[node 1] requesting goroutine profile... received response... writing binary output: debug/nodes/1/stacks.pprof... done
          -[node 1] requesting heap profile... received response... writing binary output: debug/nodes/1/heap.pprof... done
          -[node 1] requesting engine stats... received response... writing binary output: debug/nodes/1/lsm.txt... done
          -[node 1] requesting heap profile list... received response... done
          -[node ?] ? heap profiles found
          -[node 1] requesting goroutine dump list... received response... done
          -[node ?] ? goroutine dumps found
          -[node 1] requesting cpu profile list... received response... done
          -[node ?] ? cpu profiles found
          -[node 1] requesting execution trace list... received response... done
          -[node ?] ? execution traces found
          -[node 1] requesting log files list... received response... done
          -[node ?] ? log files found
          -[node 1] requesting ranges... received response... done
          -[node 1] writing ranges... writing JSON output: debug/nodes/1/ranges.json... done
          -[cluster] pprof summary script... writing binary output: debug/pprof-summary.sh... done
          -[cluster] hot range summary script... writing binary output: debug/hot-ranges.sh... done
          -[cluster] tenant hot range summary script... writing binary output: debug/hot-ranges-tenant.sh... done
          -[cluster] capture debug zip flags... writing binary output: debug/debug_zip_command_flags.txt... done
          +[cluster] requesting nodes... received response... writing JSON output: debug/nodes.json... error:json: unsupported value: +Inf
          +ERROR: json: unsupported value: +Inf
           
    panic.go:615: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/0eff02e3d605640c0624f6bcc67cf8d4/logTestZipQueryFallback905151987
--- FAIL: TestZipQueryFallback (46.06s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/obs-prs @cockroachdb/server @cockroachdb/obs-india-prs
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestZipQueryFallback.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-60549

--- a/pkg/util/metric/histogram_snapshot.go
+++ b/pkg/util/metric/histogram_snapshot.go
@@ -62,8 +62,19 @@ func (hs HistogramSnapshot) ValueAtQuantile(q float64) float64 {
 		count -= *buckets[b-1].CumulativeCount
 		rank -= *buckets[b-1].CumulativeCount
 	}
-	val := bucketStart + (bucketEnd-bucketStart)*(float64(rank)/float64(count))
-	if math.IsNaN(val) || math.IsInf(val, -1) {
+	var val float64
+	if count == 0 {
+		// The bucket has no observations. Use the bucket start as the estimate
+		// when rank is also 0 (quantile 0), otherwise use the bucket end.
+		if rank == 0 {
+			val = bucketStart
+		} else {
+			val = bucketEnd
+		}
+	} else {
+		val = bucketStart + (bucketEnd-bucketStart)*(float64(rank)/float64(count))
+	}
+	if math.IsNaN(val) || math.IsInf(val, 0) {
 		return 0
 	}
 

--- a/pkg/util/metric/metric_test.go
+++ b/pkg/util/metric/metric_test.go
@@ -370,6 +370,36 @@ func TestManualWindowHistogram(t *testing.T) {
 	}
 }
 
+// TestValueAtQuantileWithEmptyBuckets verifies that ValueAtQuantile does not
+// return +Inf when adjacent buckets have the same cumulative count (causing a
+// division by zero in the interpolation). This was the root cause of #163958.
+func TestValueAtQuantileWithEmptyBuckets(t *testing.T) {
+	u := func(v uint64) *uint64 { return &v }
+	f := func(v float64) *float64 { return &v }
+
+	// Construct a histogram where the first two buckets have the same
+	// cumulative count, meaning the second bucket is empty. This causes
+	// count to be 0 in the interpolation, producing +Inf via division by zero.
+	h := &prometheusgo.Histogram{
+		SampleCount: u(5),
+		SampleSum:   f(250),
+		Bucket: []*prometheusgo.Bucket{
+			{CumulativeCount: u(0), UpperBound: f(1)},
+			{CumulativeCount: u(0), UpperBound: f(5)},
+			{CumulativeCount: u(2), UpperBound: f(10)},
+			{CumulativeCount: u(3), UpperBound: f(25)},
+			{CumulativeCount: u(5), UpperBound: f(100)},
+		},
+	}
+
+	snap := MakeHistogramSnapshot(h)
+	for q := 0.0; q <= 100.0; q += 10 {
+		val := snap.ValueAtQuantile(q)
+		require.False(t, math.IsInf(val, 0), "ValueAtQuantile(%v) returned infinity", q)
+		require.False(t, math.IsNaN(val), "ValueAtQuantile(%v) returned NaN", q)
+	}
+}
+
 func TestManualWindowHistogramTicker(t *testing.T) {
 	now := time.UnixMicro(1699565116)
 	defer TestingSetNow(func() time.Time {

